planner/core/logical_plans.go

```diff
@@ -2054,21 +2056,53 @@ func (fb *FrameBound) Clone() *FrameBound {
 	return cloned
 }
 
-// InitCompareCols will init CompareCols
-func (fb *FrameBound) InitCompareCols(ctx sessionctx.Context, orderByCols []*expression.Column) {
+func (fb *FrameBound) updateCmpFuncsAndCmpDataType(cmpDataType types.EvalType) {
+	// When cmpDataType can't match to any condition, we can ignore it.
+	//
+	// For example:
+	//   `create table test.range_test(p int not null,o text not null,v int not null);`
+	//   `select *, first_value(v) over (partition by p order by o) as a from range_test;`
+	//   The sql's frame type is range, but the cmpDataType is ETString and when the user explicitly use range frame
+	//   the sql will raise error before generating logical plan, so it's ok to ignore it.
+	switch cmpDataType {
+	case types.ETInt:
+		fb.CmpFuncs[0] = expression.CompareInt
+		fb.CmpDataType = tipb.RangeCmpDataType_Int
+	case types.ETDatetime, types.ETTimestamp:
+		fb.CmpFuncs[0] = expression.CompareTime
+		fb.CmpDataType = tipb.RangeCmpDataType_DateTime
+	case types.ETDuration:
+		fb.CmpFuncs[0] = expression.CompareDuration
+		fb.CmpDataType = tipb.RangeCmpDataType_Duration
+	case types.ETReal:
+		fb.CmpFuncs[0] = expression.CompareReal
+		fb.CmpDataType = tipb.RangeCmpDataType_Float
+	case types.ETDecimal:
+		fb.CmpFuncs[0] = expression.CompareDecimal
+		fb.CmpDataType = tipb.RangeCmpDataType_Decimal
+	}
+}
+
+// UpdateCompareCols will update CompareCols.
+func (fb *FrameBound) UpdateCompareCols(ctx sessionctx.Context, orderByCols []*expression.Column) error {
 	if len(fb.CalcFuncs) > 0 {
```

Comment:
Return error if `len(fb.CalcFuncs) > 1`?

Reply:
> Return error if `len(fb.CalcFuncs) > 1`?

When sql is select first_value(v) over (order by o1, o2); it's frame type is range and len(fb.CalcFuncs) may be greater than 1.